refactor(collapsible): follow current style guides and fix typescript issues

components is not currently following our current style guides and tsc is returning following errors

```
21   readonly detail: StateChangedEventDetail
                      ~~~~~~~~~~~~~~~~~~~~~~~

Collapsible.tsx(21,20): error TS4033: Property 'detail' of exported interface has or is using private name 'StateChangedEventDetail'.


24 export class Collapsible extends Component<CollapsibleProps> {
                                              ~~~~~~~~~~~~~~~~

Collapsible.tsx(24,44): error TS4020: Extends clause of exported class 'Collapsible' has or is using private name 'CollapsibleProps'.


26   static get props() {
                ~~~~~

Collapsible.tsx(26,14): error TS4038: Return type of public static property getter from exported class has or is using name 'PropOptions' from external module "/Users/mario/ngParty/blaze-elements/node_modules/skatejs/src/ts-typings/api" but cannot be named.
```

--- a/packages/collapsible/Collapsible.test.tsx
+++ b/packages/collapsible/Collapsible.test.tsx
@@ -1,83 +1,138 @@
-import { Collapsible } from './Collapsible';
-
+import { Collapsible } from './index';
+import * as expect from 'expect';
 import { h, mount } from 'bore';
 import { emit } from 'skatejs';
 
 describe( Collapsible.is, () => {
 
-  it( 'should render closed', () =>
+  describe( `Custom element`, () => {
 
-    mount(
-      <Collapsible>
-        <span slot="title">Some <strong>Name</strong></span>
-        <div>
-          Some first content <strong>!!!</strong>
-        </div>
-      </Collapsible>
-    ).wait(( element ) => !element.has( 'div.c-card__item' ) ) );
+    it( `should be registered`, () => {
 
-  it( 'should open when isOpened present', () =>
+      expect( customElements.get( Collapsible.is ) ).toBe( Collapsible );
 
-    mount(
-      <Collapsible isOpened>
-        <span slot="title">Some <strong>Name</strong></span>
-        <div>
-          Some first content <strong>!!!</strong>
-        </div>
-      </Collapsible>
-    ).wait(( element ) => element.has( 'div.c-card__item' ) ) );
+    } );
 
-  it( 'should open proper collapsible', ( done ) => {
+    it( `should render via JSX IntrinsicElement`, () => {
 
-    mount(
-      <div>
-        <Collapsible id="firstCollapsible">
-          <span slot="title">Some <strong>Name</strong></span>
-          <div>
-            Some first content <strong>!!!</strong>
-          </div>
-        </Collapsible>
-        <Collapsible id="secondCollapsible">
-          <span slot="title">Some <strong>Name</strong></span>
-          <div>
-            Some first content <strong>!!!</strong>
-          </div>
-        </Collapsible>
-      </div>
-    ).waitFor(( element ) =>
-      (
-        !!element.one( '#firstCollapsible' ).node.shadowRoot &&
-        !!element.one( '#secondCollapsible' ).node.shadowRoot
-      ) ).then(( element ) => {
+      return mount(
+        <bl-collapsible />
+      ).wait(( element ) => {
 
-        const firstCollapsible = element.one( '#firstCollapsible' );
-        const secondCollapsible = element.one( '#secondCollapsible' );
-
-        expect( firstCollapsible.has( 'div.c-card__item' ) ).to.be.false;
-        expect( secondCollapsible.has( 'div.c-card__item' ) ).to.be.false;
-
-        secondCollapsible.node.addEventListener( 'stateChanged', function() {
-
-          setTimeout( function() {
-
-            expect( firstCollapsible.has( 'div.c-card__item' ) ).to.be.false;
-            expect( secondCollapsible.has( 'div.c-card__item' ) ).to.be.true;
-
-            done();
-
-          } );
-
-        } );
-
-        emit(
-          secondCollapsible.one( 'label.c-card__item' ).node,
-          'click',
-          { composed: false }
-        );
+        expect( element.node.localName ).toBe( Collapsible.is );
 
       } );
 
+    } );
+
+    it( `should render via JSX class`, () => {
+
+      return mount(
+        <bl-collapsible />
+      ).wait(( element ) => {
+
+        expect( element.has( '.c-card--accordion' ) ).toBe( true );
+
+      } );
+
+    } );
   } );
 
+  describe( `API`, () => {
+
+    describe( `[isOpened]`, () => {
+
+      it( 'should render closed', () => {
+
+        return mount(
+          <bl-collapsible>
+            <span slot="title">Some <strong>Name</strong></span>
+            <div>
+              Some first content <strong>!!!</strong>
+            </div>
+          </bl-collapsible>
+        ).wait(( element ) => !element.has( 'div.c-card__item' ) );
+
+      } );
+
+      it( 'should open when isOpened present', () => {
+
+        return mount(
+          <bl-collapsible isOpened>
+            <span slot="title">Some <strong>Name</strong></span>
+            <div>
+              Some first content <strong>!!!</strong>
+            </div>
+          </bl-collapsible>
+        ).wait(( element ) => element.has( 'div.c-card__item' ) );
+
+      } );
+
+    } );
+
+    describe( `[isLast]`, () => {
+      // TODO
+    } );
+    describe( `[isNotStandAlone]`, () => {
+      // TODO
+    } );
+
+  } );
+
+  describe( `Interactions`, () => {
+
+    it( 'should open proper collapsible', ( done ) => {
+
+      mount(
+        <div>
+          <bl-collapsible id="firstCollapsible">
+            <span slot="title">Some <strong>Name</strong></span>
+            <div>
+              Some first content <strong>!!!</strong>
+            </div>
+          </bl-collapsible>
+          <bl-collapsible id="secondCollapsible">
+            <span slot="title">Some <strong>Name</strong></span>
+            <div>
+              Some first content <strong>!!!</strong>
+            </div>
+          </bl-collapsible>
+        </div>
+      ).waitFor(( element ) =>
+        (
+          !!element.one( '#firstCollapsible' ).node.shadowRoot &&
+          !!element.one( '#secondCollapsible' ).node.shadowRoot
+        ) ).then(( element ) => {
+
+          const firstCollapsible = element.one( '#firstCollapsible' );
+          const secondCollapsible = element.one( '#secondCollapsible' );
+
+          expect( firstCollapsible.has( 'div.c-card__item' ) ).toBe( false );
+          expect( secondCollapsible.has( 'div.c-card__item' ) ).toBe( false );
+
+          secondCollapsible.node.addEventListener( 'stateChanged', () => {
+
+            setTimeout(() => {
+
+              expect( firstCollapsible.has( 'div.c-card__item' ) ).toBe( false );
+              expect( secondCollapsible.has( 'div.c-card__item' ) ).toBe( true );
+
+              done();
+
+            } );
+
+          } );
+
+          emit(
+            secondCollapsible.one( 'label.c-card__item' ).node,
+            'click',
+            { composed: false }
+          );
+
+        } );
+
+    } );
+
+  } );
 
 } );

--- a/packages/collapsible/Collapsible.tsx
+++ b/packages/collapsible/Collapsible.tsx
@@ -114,7 +114,7 @@ export default class Collapsible extends Component<CollapsibleProps> {
 
     props( this, { isOpened: !this.isOpened } );
 
-    // TODO: we should return boolean instead
+    // TODO: we should emit single boolean instead
     emit( this, 'stateChanged', {
       detail: {
         opened: this.isOpened,

--- a/packages/collapsible/Collapsible.tsx
+++ b/packages/collapsible/Collapsible.tsx
@@ -1,46 +1,68 @@
-import { h, Component, prop, emit } from 'skatejs';
+import { h, Component, emit, props } from 'skatejs';
 import styles from './Collapsible.scss';
-import { css } from '@blaze-elements/common';
+import { css, prop, GenericEvents, shadyCssStyles } from '@blaze-elements/common';
 
-// public
-// @FIXME this needs to be adressed other way (like Partial<HTMLDivElement)) or element which is root in renderCallback
-interface NativeDivAttrs {
-  id?: string,
-}
-interface CollapsibleProps extends NativeDivAttrs {
+export type CollapsibleProps = Props & EventProps;
+
+export type Attrs = {
+  'is-opened'?: boolean,
+};
+
+export type Props = {
   isOpened?: boolean,
   isLast?: boolean,
-  onStateChanged?: ( this: this, ev: StateChangedEvent ) => any,
-}
+  isNotStandAlone?: boolean,
+};
 
-interface StateChangedEventDetail {
+export type EventProps = {
+  onStateChanged?: GenericEvents.CustomChangeHandler<StateChangedEventDetail>
+};
+
+export type Events = {
+  stateChanged: GenericEvents.CustomChangeHandler<StateChangedEventDetail>
+};
+
+export interface StateChangedEventDetail {
   readonly opened: boolean;
   readonly collapsed: boolean;
 }
+
+// TODO: Deprecated remove when accordion is refactored
 export interface StateChangedEvent extends Event {
   readonly detail: StateChangedEventDetail
 }
 
-export class Collapsible extends Component<CollapsibleProps> {
-  static get is() { return 'bl-collapsible'; }
-  static get props() {
-    return {
-      isOpened: prop.boolean( {
-        attribute: true
-      } ),
-      isNotStandAlone: prop.boolean(),
-      isLast: prop.boolean()
-    };
-  }
+@shadyCssStyles()
+export default class Collapsible extends Component<CollapsibleProps> {
 
+  @prop( {
+    type: Boolean,
+    attribute: {
+      source: true
+    }
+  } )
   isOpened: boolean;
 
-  // internal property to be used from other components, when there is collection of collapsible items to indicate
-  // this is the last one in the collection
+  /**
+   * internal property to be used from other components,
+   * when there is collection of collapsible items to indicate
+   * this is the last one in the collection
+   */
+  @prop( {
+    type: Boolean
+  } )
   isLast: boolean;
 
-  // internal property to be used from other components, when there is collection of collapsible items
-  isNotStandAlone: boolean;
+  /**
+   * internal property to be used from other components,
+   * when there is collection of collapsible items
+   */
+  @prop( {
+    type: Boolean
+  } )
+  isNotStandAlone: boolean; // TODO: rename this to isCollection
+
+  get css() { return styles; }
 
   connectedCallback() {
     super.connectedCallback();
@@ -78,20 +100,21 @@ export class Collapsible extends Component<CollapsibleProps> {
       </div>
       : null;
 
-    return [
-      <style>{styles}</style>,
+    return (
       <div className={wrapperClassName}>
         <label className={headerClassName} onClick={this.handleStateChange}>
           <slot name="title" />
         </label>
         {content}
       </div>
-    ];
+    );
   }
 
   private handleStateChange( event: Event ) {
-    this.isOpened = !this.isOpened;
 
+    props( this, { isOpened: !this.isOpened } );
+
+    // TODO: we should return boolean instead
     emit( this, 'stateChanged', {
       detail: {
         opened: this.isOpened,
@@ -101,4 +124,7 @@ export class Collapsible extends Component<CollapsibleProps> {
   }
 }
 
-customElements.define( Collapsible.is, Collapsible );
+// TODO: Deprecated remove when accordion is refactored
+export {
+  Collapsible
+}

--- a/packages/collapsible/index.demo.tsx
+++ b/packages/collapsible/index.demo.tsx
@@ -1,13 +1,12 @@
 import { h, Component } from 'skatejs';
-import { Collapsible } from './Collapsible';
-import styles from './Collapsible.scss';
+import { Collapsible } from './index';
+import { customElement } from '@blaze-elements/common';
 
+@customElement( 'bl-collapsible-demo' )
 export class Demo extends Component<void> {
-  static get is() { return 'bl-collapsible-demo'; }
 
   renderCallback() {
-    return [
-      <style>{styles}</style>,
+    return (
       <fieldset>
         <legend>{Collapsible.is}</legend>
 
@@ -18,9 +17,6 @@ export class Demo extends Component<void> {
           </div>
         </Collapsible>
       </fieldset>
-    ];
+    );
   }
 }
-
-customElements.define( Demo.is, Demo );
-

--- a/packages/collapsible/index.ts
+++ b/packages/collapsible/index.ts
@@ -1,1 +1,19 @@
-import './Collapsible';
+
+import { customElement, GenericTypes } from '@blaze-elements/common';
+import RawCollapsible, { CollapsibleProps, Attrs, Events } from './Collapsible';
+
+const Collapsible = customElement( 'bl-collapsible' )( RawCollapsible ) as typeof RawCollapsible;
+
+export {
+  Collapsible
+};
+
+
+declare global {
+  namespace JSX {
+    interface IntrinsicElements {
+      'bl-collapsible': GenericTypes.IntrinsicCustomElement<CollapsibleProps>
+      & GenericTypes.IntrinsicBoreElement<Attrs, Events>
+    }
+  }
+}

